### PR TITLE
Add Google Vision OCR block to workflows

### PR DESCRIPTION
**Pull request by brunopicinin**
# Description

Adds a new workflow OCR block, based on Google Vision API. The block outputs the text for the whole image as well as detected language and `sv.Detections(...)` for text blocks with proper labels.

## Type of change

Please delete options that are not relevant.

-   [x] New feature (non-breaking change which adds functionality)

## How has this change been tested, please provide a testcase or example of how you tested the change?

1. Tested locally with an image of pure text, using `ocr_text_detection` mode:

CleanShot 2024-10-01 at 21 17 40@2x

CleanShot 2024-10-01 at 21 17 48@2x

2. Tested locally with a picture containing some text, using `text_detection` mode:

CleanShot 2024-10-01 at 21 25 39@2x

CleanShot 2024-10-01 at 21 25 47@2x

3. Tested locally with an image without text:

CleanShot 2024-10-01 at 21 26 43@2x

### CODE REVIEW

**Comment by brunopicinin** [comment]
I ended up using the `fullTextAnnotation` for the blocks. Doing some tests, it seems to give "more natural" results than the `textAnnotations` for the text blocks. Also, it comes with `confidence` score on `DOCUMENT_TEXT_DETECTION` mode.

As an exemple, the difference in the Google Vision API output for the following image:

sign

CleanShot 2024-10-01 at 21 38 39@2x

File: inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            for block in page["blocks"]:
+                # Get bounding box coordinates
+                box = block["boundingBox"]["vertices"]
+                x_min = min(v["x"] for v in box)
+                y_min = min(v["y"] for v in box)
+                x_max = max(v["x"] for v in box)
+                y_max = max(v["y"] for v in box)
+                xyxy.append([x_min, y_min, x_max, y_max])
+
+                # Only DOCUMENT_TEXT_DETECTION provides confidence score, use 1.0 otherwise
```
Comment (PawelPeczek-Roboflow) [inline inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py:147]: good choice 👍

File: inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            "version": "v1",
+            "short_description": "Detect text in images using Google Vision API",
+            "long_description": LONG_DESCRIPTION,
+            "license": "Apache-2.0",
+            "block_type": "model",
+        },
+        protected_namespaces=(),
+    )
+    type: Literal["roboflow_core/google_vision_ocr@v1"]
+    image: Union[WorkflowImageSelector, StepOutputImageSelector]
```
Comment (PawelPeczek-Roboflow) [inline inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py:57]: Could you please add pydantic fields descriptions, for image:

```
Field(
    description="The image to run OCR",
    examples=["$inputs.image", "$steps.cropping.crops"],
)
```
That would make UI to render descriptions and our docs to have nice examples - we generate this kind of pages for each block automatically: https://inference.roboflow.com/workflows/blocks/detections_consensus/#properties

File: inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    ocr_type: Literal["text_detection", "ocr_text_detection"]
```
Comment (PawelPeczek-Roboflow) [inline inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py:58]: Apart from adding description here, you can provide:
```
Field(
        description="XXX",
        json_schema_extra={
            "values_metadata": {
                "text_detection": {"name": "Human Friendly Name", "description": "Short description"},
                "ocr_text_detection": {...},
             },
        },
    )
```
That would make this change in UI:
image
In descriptions I would explain the difference posted in google docs:
* `TEXT_DETECTION detects and extracts text from any image` - in a sense - "wild" scenes depicted in images
* `DOCUMENT_TEXT_DETECTION` is optimised for documents

File: inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        # Check for image without text
+        if "textAnnotations" not in result or not result["textAnnotations"]:
+            return {
+                "text": "",
+                "language": "",
+                "predictions": sv.Detections.empty(),
+            }
+
+        # Extract predictions from the response
+        text = result["textAnnotations"][0]["description"]
```
Comment (PawelPeczek-Roboflow) [inline inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py:129]: good navigation in output - I must admit it is not 100% clear at the first glance how to parse response 👍

File: inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            data={CLASS_NAME_DATA_FIELD: np.array(classes)},
+        )
+
+        predictions[DETECTION_ID_KEY] = np.array(detections_id)
+        predictions[PREDICTION_TYPE_KEY] = np.array(["ocr"] * len(predictions))
+        image_height, image_width = image.numpy_image.shape[:2]
+        predictions[IMAGE_DIMENSIONS_KEY] = np.array(
+            [[image_height, image_width]] * len(predictions)
+        )
+
```
Comment (PawelPeczek-Roboflow) [inline inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py:176]: if you use [attach_parents_coordinates_to_sv_detections(...)](https://github.com/roboflow/inference/blob/6981108d866be40952fd32257da693cd8457e299/inference/core/workflows/core_steps/common/utils.py#L180) we would have full set of metadata in sv.detections

**Review by PawelPeczek-Roboflow — COMMENTED** [review]
Great contribution 💪 
I am very positively surprised on the speed and quality. The comments are minor - I am willing to accept the contribution and apply myself or you can do it if you wanted.

Do you have any comments / suggestions / feature requests regarding Workflows ecosystem as a result of this contribution?

File: inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py (hunk 2 above)
Comment (brunopicinin) [reply inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py:57]: Thanks for the input! I'll make the necessary fixes. I was hoping for some guidance on the project's standards 😄.

The project was easier to contribute to than expected. The links in the issue and the docs on how to contribute / workflow docs were very helpful. 👏

Comment (PawelPeczek-Roboflow) [reply inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py:57]: cool. Let's have that PR closed soon 😄

File: inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py (hunk 5 above)
Comment (brunopicinin) [reply inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py:196]: Done 👍

File: inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py (hunk 3 above)
Comment (brunopicinin) [reply inference/core/workflows/core_steps/models/foundation/google_vision_ocr/v1.py]: Done 👍

**Review by PawelPeczek-Roboflow — APPROVED** [review]

**Pull request merged**